tests/unit_tests/sparse_blas/source/sparse_gemm_buffer.cpp

```diff
@@ -0,0 +1,236 @@
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#include <algorithm>
+#include <complex>
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+#include "oneapi/mkl.hpp"
+#include "oneapi/mkl/detail/config.hpp"
+#include "sparse_reference.hpp"
+#include "test_common.hpp"
+#include "test_helper.hpp"
+
+#include <gtest/gtest.h>
+
+extern std::vector<sycl::device *> devices;
+
+namespace {
+
+template <typename fpType, typename intType>
+int test(sycl::device *dev, intType nrows_A, intType ncols_A, intType ncols_C,
+         double density_A_matrix, oneapi::mkl::index_base index,
+         oneapi::mkl::layout dense_matrix_layout, oneapi::mkl::transpose transpose_A,
+         oneapi::mkl::transpose transpose_B, fpType alpha, fpType beta, intType ldb, intType ldc) {
+    sycl::queue main_queue(*dev, exception_handler_t());
+
+    intType int_index = (index == oneapi::mkl::index_base::zero) ? 0 : 1;
+    std::size_t opa_nrows = static_cast<std::size_t>(
+        transpose_A == oneapi::mkl::transpose::nontrans ? nrows_A : ncols_A);
+    std::size_t opa_ncols = static_cast<std::size_t>(
+        transpose_A == oneapi::mkl::transpose::nontrans ? ncols_A : nrows_A);
+
+    // Input matrix in CSR format
+    std::vector<intType> ia_host, ja_host;
+    std::vector<fpType> a_host;
+    intType nnz = generate_random_matrix<fpType, intType>(nrows_A, ncols_A, density_A_matrix,
+                                                          int_index, ia_host, ja_host, a_host);
+
+    // Input and output dense vectors
+    std::vector<fpType> b_host, c_host;
+    rand_matrix(b_host, dense_matrix_layout, opa_ncols, static_cast<std::size_t>(ncols_C),
+                static_cast<std::size_t>(ldb));
+    rand_matrix(c_host, dense_matrix_layout, opa_nrows, static_cast<std::size_t>(ncols_C),
+                static_cast<std::size_t>(ldc));
+    std::vector<fpType> c_ref_host(c_host);
+
+    // Shuffle ordering of column indices/values to test sortedness
+    shuffle_data(ia_host.data(), ja_host.data(), a_host.data(), static_cast<std::size_t>(nrows_A));
+
+    auto ia_buf = make_buffer(ia_host);
+    auto ja_buf = make_buffer(ja_host);
+    auto a_buf = make_buffer(a_host);
+    auto b_buf = make_buffer(b_host);
+    auto c_buf = make_buffer(c_host);
+
+    sycl::event ev_release;
+    oneapi::mkl::sparse::matrix_handle_t handle = nullptr;
+    try {
+        CALL_RT_OR_CT(oneapi::mkl::sparse::init_matrix_handle, main_queue, &handle);
+
+        CALL_RT_OR_CT(oneapi::mkl::sparse::set_csr_data, main_queue, handle, nrows_A, ncols_A, nnz,
+                      index, ia_buf, ja_buf, a_buf);
+
+        CALL_RT_OR_CT(oneapi::mkl::sparse::optimize_gemm, main_queue, transpose_A, handle);
```

Comment:
The idea of the `optimize_*` functions is that they can be used optionally to possibly speed up computations. That means we support both "non-optimized" and "optimized" versions of the calls. Our unit tests should cover calling `gemm` with or without calls to `optimize_gemm`.

Reply:
Done in https://github.com/oneapi-src/oneMKL/pull/403/commits/0441c3ee899823eb0587f729b3cc448dfe7c6a9c